src/main/java/org/testng/xml/TestNGContentHandler.java

```diff
@@ -118,23 +118,27 @@ public InputSource resolveEntity(String systemId, String publicId)
     if (publicId == null) {
       return m_redirectionAwareResolver.resolveEntity(systemId, null);
     }
-    if (Parser.isUnRecognizedPublicId(publicId)) {
-      if (RuntimeBehavior.useSecuredUrlForDtd() && isUnsecuredUrl(publicId)) {
-        throw new TestNGException(RuntimeBehavior.unsecuredUrlDocumentation());
-      }
-      return m_redirectionAwareResolver.resolveEntity(systemId, publicId);
-    }
-    m_validate = true;
-    InputStream is = loadDtdUsingClassLoader();
-    if (is != null) {
-      return new InputSource(is);
-    }
-    System.out.println(
-        "WARNING: couldn't find in classpath "
-            + publicId
-            + "\n"
-            + "Fetching it from " + Parser.HTTPS_TESTNG_DTD_URL);
-    return super.resolveEntity(systemId, Parser.HTTPS_TESTNG_DTD_URL);
+    if (Parser.isDTDDomainInternallyKnownToTestNG(publicId)) {
+      //If the hostname is known internally to TestNG then ignore the "http" (or) "https" protocol
+      // and always first try to load the DTD from the classpath.
+      m_validate = true;
+      InputStream is = loadDtdUsingClassLoader();
+      if (is != null) {
+        return new InputSource(is);
+      }
+      //If the classpath loading of DTD fails, then we try to load it from "https" TestNG site.
+      System.out.println(
+          "WARNING: couldn't find in classpath "
+              + publicId
+              + "\n"
+              + "Fetching it from " + Parser.HTTPS_TESTNG_DTD_URL);
+      return super.resolveEntity(systemId, Parser.HTTPS_TESTNG_DTD_URL);
+    }
+    //If we are here, then we don't know the host from which user is trying to load the dtd
+    if (RuntimeBehavior.useSecuredUrlForDtd() && isUnsecuredUrl(publicId)) {
+      throw new TestNGException(RuntimeBehavior.unsecuredUrlDocumentation());
+    }
+    return m_redirectionAwareResolver.resolveEntity(systemId, publicId);
```

Comment:
I'm not sure to understand the logic between `m_redirectionAwareResolver` and `super`.
I think it should be the following:
classpath -> `m_redirectionAwareResolver.resolveEntity` -> `super.resolveEntity`

Reply:
@juherr - There is some more clean up that needs to be done on this I think. 

Here's how the signature looks like of the interface

```
public interface EntityResolver {
    public abstract InputSource resolveEntity (String publicId,
                                               String systemId)
        throws SAXException, IOException;
}
```

Note the order of those arguments (It expects a `publicid` first and then a `systemid` second

This is how the signature looks like in `org.xml.sax.helpers.DefaultHandler`

```
    public InputSource resolveEntity (String publicId, String systemId)
        throws IOException, SAXException
    {
        return null;
    }
```

Now this is how the signature looks like in `org.testng.xml.TestNGContentHandler`

```
  @Override
  public InputSource resolveEntity(String systemId, String publicId)
      throws SAXException, IOException {
```

Note how we have interchanged the order of the arguments ? We have flipped them to be systemid and publicid whereas the semantics of the base class and the interface say that the order should be other way around.

I will be raising another PR to clean up this discrepancy and perhaps then we should have a bit more clarity.
The `super.resolveEntity` should ideally speaking not be needed anymore because this PR changes the semantics of how we load DTD. If we see that its from TestNG domain, we try to load from classpath and if that fails, we load explicitly from `https://testng.org`

So the only place wherein the redirectionawareResolver is needed is when we are trying to load the DTD that exists in a domain that we dont know of, and where, there may have been a redirection configured even. Either ways `super.resolveEntity` doesn't do anything and just returns `null`